File: survival-graphics/src/main/java/net/survival/graphics/model/ModelRenderer.java
package net.survival.graphics.model;

import java.util.HashMap;

import net.survival.graphics.Bitmap;
import net.survival.graphics.opengl.GLFilterMode;
import net.survival.graphics.opengl.GLImmediateDrawCall;
import net.survival.graphics.opengl.GLTexture;
import net.survival.graphics.opengl.GLWrapMode;

public class ModelRenderer
{
    // TODO: Need to dispose these resources.
    private static final HashMap<String, GLTexture> textureCache = new HashMap<>();

    private ModelRenderer() {}

    public static void displayStaticModel(StaticModel model) {
        for (var i = 0; i < model.meshes.length; ++i) {
            var mesh = model.meshes[i];

            var texture = textureCache.get(mesh.texturePath);
            if (texture == null) {
                var bitmap = Bitmap.fromFile(mesh.texturePath);
                texture = new GLTexture();
                texture.beginBind()
                        .setMinFilter(GLFilterMode.NEAREST_MIPMAP_NEAREST)
                        .setMagFilter(GLFilterMode.NEAREST)
                        .setWrapS(GLWrapMode.REPEAT)
                        .setWrapT(GLWrapMode.REPEAT)
                        .setMipmapEnabled(true)
                        .setData(bitmap)
                        .endBind();
                textureCache.put(mesh.texturePath, texture);
            }

            var drawCall = GLImmediateDrawCall.beginTriangles(texture);
            drawCall.color(1.0f, 1.0f, 1.0f);
            makeStaticMeshDrawCall(drawCall, mesh);
            drawCall.end();
        }
    }

    private static void makeStaticMeshDrawCall(GLImmediateDrawCall drawCall,
            StaticModel.Mesh mesh)
    {
        for (var j = 0; j < mesh.vertexCount; ++j) {
            var baseIndex = j * 8;
            var x = mesh.vertices[baseIndex];
            var y = mesh.vertices[baseIndex + 1];
            var z = mesh.vertices[baseIndex + 2];
            var normalX = mesh.vertices[baseIndex + 3];
            var normalY = mesh.vertices[baseIndex + 4];
            var normalZ = mesh.vertices[baseIndex + 5];
            var texCoordU = mesh.vertices[baseIndex + 6];
            var texCoordV = mesh.vertices[baseIndex + 7];

            drawCall.texCoord(texCoordU, texCoordV);
            drawCall.normal(normalX, normalY, normalZ);
            drawCall.vertex(x, y, z);
        }
    }
}

File: survival-graphics/src/main/java/net/survival/graphics/model/ModelRenderer.kt
package net.survival.graphics.model

import net.survival.graphics.Bitmap
import net.survival.graphics.model.StaticModel.Mesh
import net.survival.graphics.opengl.GLFilterMode
import net.survival.graphics.opengl.GLImmediateDrawCall
import net.survival.graphics.opengl.GLTexture
import net.survival.graphics.opengl.GLWrapMode
import java.util.*

object ModelRenderer {
    // TODO: Need to dispose these resources.
    private val textureCache = HashMap<String, GLTexture>()

    @JvmStatic
    fun displayStaticModel(model: StaticModel) {
        for (i in model.meshes.indices) {
            val mesh = model.meshes[i]
            var texture = textureCache[mesh.texturePath]

            if (texture == null) {
                val bitmap = Bitmap.fromFile(mesh.texturePath)
                texture = GLTexture()
                texture.beginBind()
                    .setMinFilter(GLFilterMode.NEAREST_MIPMAP_NEAREST)
                    .setMagFilter(GLFilterMode.NEAREST)
                    .setWrapS(GLWrapMode.REPEAT)
                    .setWrapT(GLWrapMode.REPEAT)
                    .setMipmapEnabled(true)
                    .setData(bitmap)
                    .endBind()
                textureCache[mesh.texturePath] = texture
            }

            val drawCall = GLImmediateDrawCall.beginTriangles(texture)
            drawCall.color(1.0f, 1.0f, 1.0f)
            makeStaticMeshDrawCall(drawCall, mesh)
            drawCall.end()
        }
    }

    private fun makeStaticMeshDrawCall(drawCall: GLImmediateDrawCall, mesh: Mesh) {
        for (j in 0 until mesh.vertexCount) {
            val baseIndex = j * 8
            val x = mesh.vertices[baseIndex]
            val y = mesh.vertices[baseIndex + 1]
            val z = mesh.vertices[baseIndex + 2]
            val normalX = mesh.vertices[baseIndex + 3]
            val normalY = mesh.vertices[baseIndex + 4]
            val normalZ = mesh.vertices[baseIndex + 5]
            val texCoordU = mesh.vertices[baseIndex + 6]
            val texCoordV = mesh.vertices[baseIndex + 7]
            drawCall.texCoord(texCoordU, texCoordV)
            drawCall.normal(normalX, normalY, normalZ)
            drawCall.vertex(x, y, z)
        }
    }
}